### PR TITLE
529 resize token transfer cards

### LINKED ISSUE #529: Reduce the height of token transfer transaction cards
See #478 for some ideas. We need to alter the design to make the approach taken in #478 simpler.

### PR DESCRIPTION
**Pull request by katibest**
Resolves #529 
## Motivation
Reduce size of token transfer tiles by altering UI.

## Changelog
* Added an expand so not all transfers are listed on a tile initially, "view more transfers"  allows to see all of them is needed
* Styled the list of transfers 

expand-token-tile

### CODE REVIEW

File: apps/block_scout_web/assets/css/components/_token_tile_view_more.scss [hunk 1]
```diff
@@ -0,0 +1,66 @@
+.token-tile-more {
```
Comment (feliperenan) [inline apps/block_scout_web/assets/css/components/_token_tile_view_more.scss]: The file name could be `token_tile_more` as well.

File: apps/block_scout_web/assets/css/components/_token_tile_view_more.scss [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  margin-right: 15px;
+}
+.token-tile-more span:after {
+  left: 55%;
+  margin-left: 15px;
+}
+
+
+
+@media (min-width: 768px) {
```
Comment (feliperenan) [inline apps/block_scout_web/assets/css/components/_token_tile_view_more.scss]: Doesn't make have variables to our breakpoints? Something like:

```
$mobile = 375px;
$tablet  = 768px
...
```

File: apps/block_scout_web/lib/block_scout_web/templates/address_transaction/_transaction.html.eex [hunk 3]
```diff
@@ -59,41 +52,25 @@
     </div>
 
     <%= if BlockScoutWeb.TransactionView.involves_token_transfers?(@transaction) do %>
-      <div class="offset-md-2 col-md-10">
-        <hr class="mt-3 mb-3 w-100" />
-        <p class="tile-title"><%= gettext "Transfers" %></p>
-      </div>
-    <% end %>
+      <div class="offset-md-2 col-md-10 d-flex flex-column mt-1 mb-2">
+        <% [first_token_transfer | tail]= @transaction.token_transfers %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/address_transaction/_transaction.html.eex]: In this context, I would avoid `tail` name and would try to use something that represents more that last token transfers. Something like `tokens_tranfers_without_first_item` or just `token_transfers` 🤔.

File: apps/block_scout_web/lib/block_scout_web/templates/address_transaction/_transaction.html.eex [hunk 4]
```diff
@@ -59,41 +52,25 @@
...
-            </span>
-          <% end %>
-        <% end %>
+          <div class="collapse" id="<%= @transaction.hash %>">
+            <%= for token_transfer <- tail do %>
+              <%= render "_token_transfer.html", address: @address, locale: @locale, token_transfer: token_transfer %>
+            <% end %>
+          </div>
       </div>
+      <%= if length(tail) != 0 do %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/address_transaction/_transaction.html.eex]: We could use `Enum.any?/1` instead.

File: apps/block_scout_web/lib/block_scout_web/templates/address_transaction/_transaction.html.eex [hunk 5]
```diff
@@ -59,41 +52,25 @@
...
+        <div class="col-md-12 d-flex flex-column text-center token-tile-more mt-2">
+          <span>
+            <%= link to: "##{@transaction.hash}", "data-toggle": "collapse", "data-test": "token_transfers_expansion" do %>
+              <%= gettext "View More Transfers" %>
+              <% end %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/address_transaction/_transaction.html.eex]: 🎨 wrong indentation here.

File: apps/block_scout_web/lib/block_scout_web/templates/transaction/_tile.html.eex [hunk 6]
```diff
@@ -19,17 +19,41 @@
...
+      <span class="mr-2 mr-md-0 order-1">
         <%= link(
           gettext("Block #%{number}", number: to_string(@transaction.block.number)),
           to: block_path(BlockScoutWeb.Endpoint, :show, @locale, @transaction.block)
         ) %>
       </span>
       <span data-from-now="<%= @transaction.block.timestamp %>"></span>
     </div>
+
+    <%= if BlockScoutWeb.TransactionView.involves_token_transfers?(@transaction) do %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/transaction/_tile.html.eex]: It seems like the same markup that [this one](https://github.com/poanetwork/blockscout/pull/637/files#diff-902d21cb2dab9beb940dcc37aedc262dR54)

Would make sense move it to partial? 🤔

File: apps/block_scout_web/lib/block_scout_web/views/transaction_view.ex [hunk 7]
```diff
@@ -45,9 +45,8 @@ defmodule BlockScoutWeb.TransactionView do
     AddressView.contract?(from_address) || AddressView.contract?(to_address)
   end
 
-  def involves_token_transfers?(%Transaction{} = transaction) do
-    Ecto.assoc_loaded?(transaction.token_transfers) && Enum.any?(transaction.token_transfers)
-  end
+  def involves_token_transfers?(%Transaction{token_transfers: []}), do: false
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/views/transaction_view.ex]: nice ❤️

File: apps/block_scout_web/test/block_scout_web/features/viewing_addresses_test.exs [hunk 8]
```diff
@@ -298,6 +294,46 @@ defmodule BlockScoutWeb.ViewingAddressesTest do
       taft = addresses.taft
       morty = build(:address)
 
+      contract_token_address = insert(:contract_address)
```
Comment (feliperenan) [inline apps/block_scout_web/test/block_scout_web/features/viewing_addresses_test.exs]: ❤️

File: apps/block_scout_web/lib/block_scout_web/templates/address_transaction/_transaction.html.eex [hunk 9]
```diff
@@ -59,41 +52,25 @@
...
+            <%= for token_transfer <- tail do %>
+              <%= render "_token_transfer.html", address: @address, locale: @locale, token_transfer: token_transfer %>
+            <% end %>
+          </div>
       </div>
+      <%= if length(tail) != 0 do %>
+        <div class="col-md-12 d-flex flex-column text-center token-tile-more mt-2">
+          <span>
+            <%= link to: "##{@transaction.hash}", "data-toggle": "collapse", "data-test": "token_transfers_expansion" do %>
+              <%= gettext "View More Transfers" %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/address_transaction/_transaction.html.eex]: We could change the text when the collapse is opened to something like: `Hide token transfers`. Not sure if it's a good ideia 🙈. Wdyt?

File: apps/block_scout_web/lib/block_scout_web/templates/transaction/_tile.html.eex [hunk 10]
```diff
@@ -19,17 +19,41 @@
...
+              <%= render BlockScoutWeb.AddressTransactionView, "_token_transfer.html", address: assigns[:address], locale: @locale, token_transfer: token_transfer %>
+            <% end %>
+          </div>
+      </div>
+      <%= if length(tail) != 0 do %>
+        <div class="col-md-12 d-flex flex-column mt-1 mb-2 text-center token-tile-more">
+          <span>
+            <%= link to: "##{@transaction.hash}", "data-toggle": "collapse", "data-test": "token_transfers_expansion" do %>
+              <%= gettext "View More Transder" %>
+          <% end %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/transaction/_tile.html.eex]: 🎨 wrong indentation.

File: apps/block_scout_web/assets/css/components/_token_tile_view_more.scss (hunk 2 above)
Comment (katibest) [reply apps/block_scout_web/assets/css/components/_token_tile_view_more.scss]: I always forget about those guys! Thank you!

File: apps/block_scout_web/assets/css/components/_token_tile_view_more.scss [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+@media (max-width: 375px) {
+
+  .token-tile-more span:before,
+  .token-tile-more span:after {
+    width: 28%;
+  }
+  .token-tile-more span:before {
+    right: 66%;
+    margin-right: 15px;
```
Comment (feliperenan) [inline apps/block_scout_web/assets/css/components/_token_tile_view_more.scss]: I have used relative size based on the root size in the CSS that I have written. So here I would consider `.9375rem` (the result of 15/16). Is that a good idea, if so, can we use here as well?

File: apps/block_scout_web/assets/css/components/_token_tile_view_more.scss (hunk 1 above)
Comment (katibest) [reply apps/block_scout_web/assets/css/components/_token_tile_view_more.scss]: I might switch it and make the class .token_tile_view_more just so its more specific. 🤔

File: apps/block_scout_web/lib/block_scout_web/templates/address/_responsive_hash.html.eex [hunk 12]
```diff
@@ -1,4 +1,8 @@
-<span class="<%= if @contract do %>contract-address<% end %>">
-  <span class="d-none d-md-none d-lg-inline"><%= @address_hash %></span>
-  <span class="d-md-inline-block d-lg-none "><%= BlockScoutWeb.AddressView.trimmed_hash(@address_hash) %></span>
+<span class="<%= if @contract do %>contract-address<% end %>" data-address-hash="<%= @address_hash %>">
+  <%= if assigns[:truncate] do %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/address/_responsive_hash.html.eex]: I think we can use `@truncate`, can we? 🤔

File: apps/block_scout_web/assets/css/components/_token_tile_view_more.scss (hunk 1 above)
Comment (feliperenan) [reply apps/block_scout_web/assets/css/components/_token_tile_view_more.scss]: yeah, that's make sense to me too.

File: apps/block_scout_web/assets/css/components/_token_tile_view_more.scss (hunk 2 above)
Comment (katibest) [reply apps/block_scout_web/assets/css/components/_token_tile_view_more.scss]: Updated this but with bootstrap 4 variables - which is similar!

File: apps/block_scout_web/lib/block_scout_web/templates/transaction/_tile.html.eex [hunk 13]
```diff
@@ -1,35 +1,75 @@
...
+          <% else %>
+            <span data-test="transaction_type" class="badge badge-success tile-badge">
+              <%= gettext "IN" %>
+            </span>
+          <% end %>
+        </span>
+      <% end %>
     </div>
+
+    <%= if involves_token_transfers?(@transaction) do %>
```
Comment (feliperenan) [inline apps/block_scout_web/lib/block_scout_web/templates/transaction/_tile.html.eex]: ❤️ Nice, now we have it just in one place 🎉

**Comment by katibest** [comment]
@acravenho  We switched the transfer list to have to/from address first giving the amount plenty of room to be super long 🙃

File: apps/block_scout_web/assets/js/lib/token_transfers_toggle.js [hunk 14]
```diff
@@ -0,0 +1,11 @@
+import $ from 'jquery'
+
+$('.token-transfer-toggle').on('show.bs.collapse', () => {
```
Comment (feliperenan) [inline apps/block_scout_web/assets/js/lib/token_transfers_toggle.js]: This way the script always will consider the first `.token-transfer-toggle` that it finds. See:



We could add the selector above to listen to all collapse events that will happen inside the card. Something like:

On [line 2] in `tile.html.eex`:(https://github.com/poanetwork/blockscout/pull/637/commits/c6b47c96a92eed5d756fdd9d0e638e5c80dff1d3#diff-9c0443561a66a7a6f2c97882065f02c9R2)
```HTML
<div data-token-transfers-toggle class="row" data-test="chain_transaction">
...
</div>
```

And the script would be:

```JS
const tokenTransferToggle = (element) => {
  const $element = $(element)
  const $tokenTransferClose = $element.find('[data-token-transfer-close]')
  const $tokenTransferOpen  = $element.find('[data-token-transfer-open]')

  $element.on('show.bs.collapse', () => {
    debugger
    $tokenTransferOpen.addClass('d-none')
    $tokenTransferClose.removeClass('d-none')
  })

  $element.on('hide.bs.collapse', () => {
    debugger
    $tokenTransferClose.addClass('d-none')
    $tokenTransferOpen.removeClass('d-none')
  })
}

// Initialize the script scope for each card.
$('[data-token-transfers-toggle]').each((_index, element) => tokenTransferToggle(element))
```
Notice that I'm using data-attributes to search the titles.

Comment (Stamates) [reply apps/block_scout_web/assets/js/lib/token_transfers_toggle.js]: Good catch!  I just pushed before standup and didn't really look through what I did enough.

File: apps/explorer/test/explorer/token/balance_reader_test.exs [hunk 15]
```diff
@@ -62,23 +62,6 @@ defmodule Explorer.Token.BalanceReaderTest do
     )
   end
 
-  defp get_balance_from_blockchain_with_balance_zero() do
```
Comment (feliperenan) [inline apps/explorer/test/explorer/token/balance_reader_test.exs]: Ow, I forget this one 🙈

**Pull request merged**